Drop `test_steal_communication_heavy_tasks` since it contradicts test_d…

hendrikmakait opened the pull request:
This PR drops _flaky_ `test_steal_communication_heavy_tasks`_ since it contradicts `test_do_not_steal_communication_heavy_tasks` and should outdated after the latest changes to work-stealing.

- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

fjetter commented:
If they are contradicting tests, why is it flaky? Shouldn't it reliably fail? This still sounds like there is a bug somewhere

hendrikmakait commented:
The entire test makes fairly little sense at the moment. I think it should have been dropped in #7075 and yet I missed it. We _could_ spend time cleaning up the test, but the way it is written right now, I think it is prone to a bunch of race conditions and does not test anything useful. Also, any useful behavior should be tested in other places (e.g. `test_do_not_steal_communication_heavy_tasks`, `test_steal_expensive_data_slow_computation`)

crusaderky commented on distributed/tests/test_steal.py:
```diff
@@ -825,44 +825,6 @@ def block_reduce(x, y, event):
     assert not b.data
```
`test_do_not_steal_communication_heavy_tasks` doesn't look right, but maybe I'm missing something?

Firstly, there's a race condition. The test waits for x and y to enter running state on the worker, polling every 100ms. 
The first iteration of `while not a.state.tasks` will always fail (because we never yielded the event loop snce submit).
The second iteration, 100ms later, *most times* will find that a and b are in memory and two block_reduce are executing. However there's nothing guaranteeing it, so when you call `steal.balance()` on the next line you might have x or y still running, or in memory but without the scheduler knowing yet. Unlikely, but possible.

The test implicitly relies on tasks of unknown duration to be stolen (#5572). It should be changed not to rely on this specific use case.

Finally, why bother calling `steal.stop()`? If the block_reduce tasks can't be stolen, it should be inconsequential.

crusaderky replied:
Overhauled test in #7250. This PR can be merged as is.

crusaderky commented:
> If they are contradicting tests, why is it flaky? Shouldn't it reliably fail? This still sounds like there is a bug somewhere

It doesn't fail because it doesn't actually test where the tasks and up being exeecuted.

hendrikmakait commented:
> Unsure why it's flaky - @hendrikmakait any insight? Do you have a log?

I've run it a couple thousand times locally without being able to reproduce the flake. Since the implementation of `test_steal_communication_heavy_tasks` does not do anything useful (and is weird), I would not want to spend more time debugging it. 

> `test_do_not_steal_communication_heavy_tasks` doesn't look right.

You have a good point there, I have dropped that one as well and added three other tests that should cover the functionality that we initially wanted to test here. I like your use of `wait_for_state` in #7250 and will create a follow-up PR  that introduces that to the `assert_balance` and `_run_dependency_test` helpers to get rid of the awkward looping. 

> Finally, why bother calling `steal.stop()`? If the `block_reduce` tasks can't be stolen, it should be inconsequential.

`steal.stop()` functions as a glorified `wait()` that blocks until all stealing requests have been dealt with and are no longer `in_flight`.

crusaderky commented on distributed/tests/test_steal.py:
```diff
@@ -1446,6 +1446,57 @@ def func(*args):
 
 
+def test_balance_steal_communication_heavy_tasks():
+    dependencies = {"a": 10, "b": 10}
```
If I change this line to `{"a": 1e-6, "b": 1e-6}` the test remains green, and if go lower than that I instead get "ValueError: Expected a value larger than 16 integer but got 10."
 Are we sure we're actually testing anything here?

hendrikmakait replied:
This test breaks if we double the cost of the dependencies to `{"a": 20, "b": 20}` as the tasks have become _too_ expensive to move in this setup. By reducing the cost of the dependencies, we are checking whether we would move tasks that are cheap to move which is the desired behavior. Reducing the cost below `1e-6` creates trouble since we are calculating the actual size of the dependencies as the product of their specified cost and the available bandwidth.

crusaderky commented:
I think I was too hasty in approving.
1. See comment above
1. You forgot to actually delete `test_steal_communication_heavy_tasks`
2. Isn't `test_do_not_steal_communication_heavy_tasks` redundant now?

hendrikmakait commented:
> 2. You forgot to actually delete `test_steal_communication_heavy_tasks`
> 3. Isn't `test_do_not_steal_communication_heavy_tasks` redundant now?

Yikes, looks like merging `main` went wrong and brought those back in, sorry for missing that! See #7269 for a follow-up that fixes that.
